lib/events/events.go

```diff
@@ -351,3 +353,54 @@ func Error(err error) *string {
 	str := err.Error()
 	return &str
 }
+
+// Wait for the given event to become available, or for a timeout, whichever is sooner.
+// Returns true if the event became available, or false if we timed out.
+func (s *bufferedSubscription) waitForNewEventWithTimeout(id int, timeout time.Duration) bool {
+	// Checks whether 'id' is available. If it is, returns nil. If it isn't, returns newEventNotification (constructing it if necessary)
+	checkIDAndConstructNotificationIfNecessary := func() <-chan struct{} {
```

Comment:
I'm not super fond of this construction but I can't exactly pinpoint the problem.

Really I think I'd want this whole thing moved out into `sync.TimeoutCond` that is like a `sync.Cond` with the behavior we need...